sklearn/_fast_gradient_boosting/_loss.pyx

```diff
@@ -0,0 +1,134 @@
+        const Y_DTYPE_C [::1] y_true,
+        const Y_DTYPE_C [::1] raw_predictions):
+    cdef:
+        int n_samples
+        int i
+
+    n_samples = raw_predictions.shape[0]
+    for i in prange(n_samples, schedule='static', nogil=True):
+        # Note: a more correct exp is 2 * (raw_predictions - y_true) but
+        # since we use 1 for the constant hessian value (and not 2) this
+        # is strictly equivalent for the leaves values.
+        gradients[i] = raw_predictions[i] - y_true[i]
+
+
+cdef void _update_gradients_hessians_binary_crossentropy_parallel(
+        G_H_DTYPE_C [::1] gradients,
+        G_H_DTYPE_C [::1] hessians,
+        const Y_DTYPE_C [::1] y_true,
+        const Y_DTYPE_C [::1] raw_predictions):
+    cdef:
+        int n_samples
+        Y_DTYPE_C p_i  # proba that ith sample belongs to positive class
+        int i
+
+    n_samples = raw_predictions.shape[0]
+    for i in prange(n_samples, schedule='static', nogil=True):
+        p_i = _cexpit(raw_predictions[i])
+        gradients[i] = p_i - y_true[i]
+        hessians[i] = p_i * (1. - p_i)
+
+
+cdef void _update_gradients_hessians_categorical_crossentropy_parallel(
+        G_H_DTYPE_C [:, ::1] gradients,  # shape (pred_dim, n_samples), OUT
+        G_H_DTYPE_C [:, ::1] hessians,  # shape (pred_dim, n_samples), OUT
+        const Y_DTYPE_C [::1] y_true,  # shape (n_samples,), IN
+        # shape (pred_dim, n_samples), IN
+        const Y_DTYPE_C [:, ::1] raw_predictions):
+    cdef:
+        int prediction_dim = raw_predictions.shape[0]
+        int n_samples = raw_predictions.shape[1]
+        int k  # class index
+        int i  # sample index
+        # p[i, k] is the probability that class(ith sample) == k.
+        # It's the softmax of the raw predictions
+        Y_DTYPE_C [:, ::1] p = np.empty(shape=(n_samples, prediction_dim))
+        Y_DTYPE_C p_i_k
+
+    for i in prange(n_samples, schedule='static', nogil=True):
+        # first compute softmaxes of sample i for each class
+        for k in range(prediction_dim):
+            p[i, k] = raw_predictions[k, i]  # prepare softmax
+        _compute_softmax(p, i)
+        # then update gradients and hessians
+        for k in range(prediction_dim):
+            p_i_k = p[i, k]
+            gradients[k, i] = p_i_k - (y_true[i] == k)
+            hessians[k, i] = p_i_k * (1. - p_i_k)
+
+
+cdef inline void _compute_softmax(Y_DTYPE_C [:, ::1] p, const int i) nogil:
+    """Compute softmaxes of values in p[i, :]."""
+    # i needs to be passed (and stays constant) because otherwise Cython does
+    # not generate optimal code
+
+    cdef:
+        Y_DTYPE_C max_value = p[i, 0]
+        Y_DTYPE_C sum_exps = 0.
+        unsigned int k
+        unsigned prediction_dim = p.shape[1]
+
+    # Compute max value of array for numerical stability
+    for k in range(1, prediction_dim):
+        if max_value < p[i, k]:
+            max_value = p[i, k]
+
+    for k in range(prediction_dim):
```

Comment:
I wonder if all these for loops would be better off with numpy vector operations.

Reply:
Everything I tried with numpy would create Python interactions which are impossible without the gil :/